crates/ruma-identifiers/src/client_secret.rs

```diff
@@ -0,0 +1,172 @@
+//! Client secret identifier.
+
+use std::{convert::TryFrom, fmt, mem, rc::Rc, str::FromStr, sync::Arc};
+
+use ruma_identifiers_validation::client_secret::validate;
+
+/// A client secret.
+///
+/// Session IDs in Matrix are opaque character sequences of `[0-9a-zA-Z.=_-]. Their length must
+/// must not exceed 255 characters.
+#[repr(transparent)]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize), serde(transparent, crate = "serde"))]
+pub struct ClientSecret(str);
+
+/// An owned client secret.
+pub type ClientSecretBox = Box<ClientSecret>;
+
+impl ClientSecret {
+    #[allow(clippy::transmute_ptr_to_ptr)]
+    fn from_borrowed(s: &str) -> &Self {
+        unsafe { mem::transmute(s) }
+    }
+
+    fn from_owned(s: Box<str>) -> Box<Self> {
+        unsafe { mem::transmute(s) }
```

Comment:
This should use `Box::from_raw` + `Box::into_raw` like the `opaque_identifier` macro. Seems like I had missed to update `server_name.rs` last time.

Reply:
Would it be a good idea to make `ServerName` be defined using a macro too?

Reply:
Yes!